src/HyperlaneArbiter.sol

```diff
@@ -124,13 +124,59 @@ contract HyperlaneArbiter is Router {
         );
     }
 
+    /**
+     * @notice Quotes a compact intent and returns the msg.value that should be provided
+     * cover all hyperlane fees (relay, etc).
+     * @param claimChain The chain ID of the claim.
+     * @param compact The compact intent to fill.
+     * @dev signatures must be compliant with https://eips.ethereum.org/EIPS/eip-2098
+     * @param allocatorSignature The allocator's signature.
+     * @param sponsorSignature The sponsor's signature.
+     * @return The quoted fee.
+     */
+    function quote(
+        uint32 claimChain,
+        Compact calldata compact,
+        Intent calldata intent,
+        bytes calldata allocatorSignature,
+        bytes calldata sponsorSignature
+    ) external view returns (uint256) {
+        require(block.chainid == intent.chainId, "invalid chain");
+
+        // TODO: support Permit2 fills
+        address filler = msg.sender;
+
+        return _Router_quoteDispatch(
+            claimChain,
+            Message.encode(compact, allocatorSignature, sponsorSignature, hash(intent), intent.fee, filler),
+            "",
+            address(hook)
+        );
+    }
+
     function hash(Intent memory intent) public pure returns (bytes32) {
         return
             keccak256(abi.encode(TYPEHASH, intent.fee, intent.chainId, intent.token, intent.recipient, intent.amount));
     }
 
-    function getCompactWitnessTypestring() external pure returns (string memory) {
-        return WITNESS_TYPESTRING;
+    function getCompactWitnessDetails()
+        external
+        pure
+        returns (
+            string memory typestring,
+            string[] memory compactTokenArguments,
+            string[3][] memory customTokenArguments
+        )
+    {
+        typestring = WITNESS_TYPESTRING;
+
+        // Arguments that refer to some amount of the token in the resource lock
+        compactTokenArguments = new string[](1);
+        compactTokenArguments[0] = "fee";
+
+        // Arguments that refer to some amount of an arbitrary chain + token combination
+        customTokenArguments = new string[3][](1);
+        customTokenArguments[0] = ["intent.chainId", "intent.token", "intent.amount"];
```

Comment:
would be very happy to just adopt whatever standard struct you propose on the compact
this would facilitate both users and fillers to understand the eip712 payloads better

Reply:
I do feel like some form of JSON payload that lays out the typestring as well as the various semantic context and relations between the arguments could be a better way than .. whatever this is 🤣 

that said, it does get the job done for just detecting token names / decimals to use 🤷